src/boot/fwupd.c

```diff
@@ -0,0 +1,567 @@
+                .attrs = flags,
+                .size = size,
+                .info = info,
+        };
+        TAKE_PTR(infop);
+
+        log_debug("Parsed UEFI update variable '%ls': flags=0x%08x, hw_inst=%" PRIu64 ", status=%" PRIu32,
+                  name,
+                  info->capsule_flags,
+                  info->hw_inst,
+                  info->status);
+
+        return EFI_SUCCESS;
+}
+
+static EFI_STATUS parse_update_vars(FwupUpdateTable ***updates, size_t *n_updates) {
+        size_t name_length = 1024 * sizeof(char16_t);
+        _cleanup_free_ char16_t *name = NULL;
+        EFI_STATUS err;
+
+        assert(updates);
+        assert(n_updates);
+
+        log_debug("Searching for UEFI updates...");
+
+        /* A series of fwupd-* named variables will be prepared by userspace, but there is no fixed name,
+         * so we need to iterate over the full table of variables and find them. Unfortunately there are no
+         * APIs to search by GUID, so it's really the entire table that has to be walked. */
+
+        name = xnew0(char16_t, name_length);
+
+        for (;;) {
+                EFI_GUID guid;
+
+                err = RT->GetNextVariableName(&name_length, name, &guid);
+                if (err == EFI_NOT_FOUND)
+                        break;
+                if (err == EFI_BUFFER_TOO_SMALL) {
+                        if (name_length > SIZE_MAX / 2)
+                                return log_error_status(EFI_OUT_OF_RESOURCES, "Variable name too long: %zu", name_length);
+                        name_length *= 2;
+                        name = xrealloc(name, name_length / 2, name_length);
+                        continue;
+                }
+                if (err != EFI_SUCCESS)
+                        return log_error_status(err, "Failed to get next variable name: %m");
+
+                if (!efi_guid_equal(&guid, MAKE_GUID_PTR(FWUPDATE)))
+                        continue;
+                if (!startswith_no_case(name, u"fwupd-"))
+                        continue;
+
+                log_debug("Found UEFI update variable '%ls'", name);
+
+                _cleanup_(fwup_update_table_freep) FwupUpdateTable *update = NULL;
+                err = parse_info_var(name, &update);
+                if (err != EFI_SUCCESS) {
+                        (void) efivar_unset(MAKE_GUID_PTR(FWUPDATE), name, 0);
+                        return err;
+                }
+                if (!FLAGS_SET(update->info->status, FWUPDATE_ATTEMPT_UPDATE))
+                        continue;
+
+                EFI_TIME now;
+                err = RT->GetTime(&now, NULL);
+                if (err == EFI_SUCCESS)
+                        update->info->time_attempted = now;
+                update->info->status = FWUPDATE_ATTEMPTED;
+                *updates = xrealloc(*updates,
+                                    *n_updates * sizeof(FwupUpdateTable),
+                                    (*n_updates + 1) * sizeof(FwupUpdateTable));
```

Comment:
overflow check would be good

Reply:
glad we'll be safe on a system that has 18446744073709551615 efi variables configured :-)